File: iOSObjectiveCPlugin/Classes/Frontend/Notify/FCMNotifierPlugin.m
// ======================================================================
// Project Name    : ios_plugin
//
// Copyright © 2019 U-CREATES. All rights reserved.
//
// This source code is the property of U-CREATES.
// If such findings are accepted at any time.
// We hope the tips and helpful in developing.
// ======================================================================
#import <UIKit/UIKit.h>
#import <UserNotifications/UserNotifications.h>
#import <FirebaseInstanceID/FirebaseInstanceID.h>
#import <FirebaseCore/FirebaseCore.h>
#import <FirebaseMessaging/FirebaseMessaging.h>
#import "FCMNotifierPlugin.h"
@interface FCMNotifierPlugin ()<UNUserNotificationCenterDelegate, FIRMessagingDelegate>
@end
@implementation FCMNotifierPlugin
+ (FCMNotifierPlugin*)getInstance {
    static FCMNotifierPlugin* instance = nil;
    if (nil == instance) {
        instance = [[FCMNotifierPlugin alloc] init];
    }
    return instance;
}
- (void)regist {
    [FIRApp configure];
    FIRMessaging* messaging = [FIRMessaging messaging];
    NSNotificationCenter* notificationCenter = [NSNotificationCenter defaultCenter];
    messaging.delegate = self;
    [notificationCenter addObserver:self selector:@selector(tokenRefreshNotification) name:kFIRInstanceIDTokenRefreshNotification object:nil];
    typedef void (^cb_t)(BOOL granted, NSError * __nullable error);
    cb_t cb = ^(BOOL granted, NSError * __nullable error) {
        if (NO == granted) {
            return;
        }
        if (nil != error) {
            NSLog(@"%@", error.debugDescription);
        } else {
            FIRInstanceID* firInstance = [FIRInstanceID instanceID];
            NSString* fcmToken = [firInstance token];
            UIApplication* app = [UIApplication sharedApplication];
            [app registerForRemoteNotifications];
            NSLog(@"fcmToken::%@", fcmToken);
        }
        return;
    };
    UNAuthorizationOptions options = UNAuthorizationOptionBadge | UNAuthorizationOptionSound | UNAuthorizationOptionAlert;
    UNUserNotificationCenter* userNotoficationCenter = UNUserNotificationCenter.currentNotificationCenter;
    userNotoficationCenter.delegate = self;
    [userNotoficationCenter requestAuthorizationWithOptions:options completionHandler:cb];
    return;
}
- (void)notify:(NSString*)title body:(NSString*)body interval:(NSTimeInterval)interval {
    return;
}
- (void)notify:(NSDictionary*)userInfo {
    return;
}
- (void)reset {
    UIApplication* app = [UIApplication sharedApplication];
    app.applicationIconBadgeNumber = 0;
    return;
}
- (void)tokenRefreshNotification {
    FIRInstanceID* firInstance = [FIRInstanceID instanceID];
    FIRMessaging* messaging = [FIRMessaging messaging];
    NSString* fcmToken = [firInstance token];
    if (nil == fcmToken) {
        NSLog(@"fcmToken is null");
        return;
    }
    typedef void (^cb_t)(NSError * __nullable error);
    cb_t cb = ^(NSError * __nullable error) {
        if (nil != error) {
            NSLog(@"%@", error.debugDescription);
        } else {
            NSLog(@"success connect to fcm");
        }
        return;
    };
    [messaging disconnect];
    [messaging connectWithCompletion:cb];
    NSLog(@"fcmToken::%@", fcmToken);
    return;
}
- (void)applicationReceivedRemoteMessage:(nonnull FIRMessagingRemoteMessage*)remoteMessage {
    NSLog(@"Received data message:%@", remoteMessage.appData);
    return;
}
- (void)userNotificationCenter:(UNUserNotificationCenter*)center willPresentNotification:(UNNotification*)notification withCompletionHandler:(void (^)(UNNotificationPresentationOptions options))completionHandler __IOS_AVAILABLE(10.0)__TVOS_AVAILABLE(10.0)__WATCHOS_AVAILABLE(3.0) {
    NSDictionary* userInfo = notification.request.content.userInfo;
    [self notify:userInfo];
    UNNotificationPresentationOptions options = UNNotificationPresentationOptionSound | UNNotificationPresentationOptionBadge | UNNotificationPresentationOptionAlert;
    completionHandler(options);
    return;
}
- (void)userNotificationCenter:(UNUserNotificationCenter*)center didReceiveNotificationResponse:(UNNotificationResponse*)response withCompletionHandler:(void (^)(void))completionHandler __IOS_AVAILABLE(10.0)__WATCHOS_AVAILABLE(3.0)__TVOS_PROHIBITED {
    NSDictionary* userInfo = response.notification.request.content.userInfo;
    [self notify:userInfo];
    completionHandler();
    return;
}
@end

File: iOSSwiftPlugin/Classes/Frontend/Notify/FCMNotifierPlugin.swift
// ======================================================================
// Project Name    : ios_foundation
//
// Copyright © 2019 U-CREATES. All rights reserved.
//
// This source code is the property of U-CREATES.
// If such findings are accepted at any time.
// We hope the tips and helpful in developing.
// ======================================================================
import Firebase
import FirebaseInstanceID
import Foundation
import UIKit
import UserNotifications
public class FCMNotifierPlugin: NSObject, UNUserNotificationCenterDelegate, MessagingDelegate {
    fileprivate static var instance: FCMNotifierPlugin?
    public override init() {}
    public static func getInstance() -> FCMNotifierPlugin! {
        if (FCMNotifierPlugin.instance == nil) {
            FCMNotifierPlugin.instance = FCMNotifierPlugin()
        }
        return FCMNotifierPlugin.instance!
    }
    open func regist() -> Void {
        FirebaseApp.configure()
        let messaging: Messaging = Messaging.messaging()
        let notificationCenter: NotificationCenter = NotificationCenter.default
        messaging.delegate = self
        notificationCenter.addObserver(self, selector: #selector(self.tokenRefreshNotification), name: NSNotification.Name.InstanceIDTokenRefresh, object: nil)
        if #available(iOS 10.0, *) {
            func callback(granted: Bool, error: Error?) -> Void {
                if (granted == false) {
                    return
                }
                if (error != nil) {
                    NSLog(error.debugDescription)
                } else {
                    let firInstance: InstanceID = InstanceID.instanceID()
                    let fcmToken: String? = firInstance.token()
                    let app: UIApplication = UIApplication.shared
                    app.registerForRemoteNotifications()
                    NSLog(String(format: "fcmToken::%@", arguments: [fcmToken!]))
                }
                return
            }
            let userNotificationCenter: UNUserNotificationCenter = UNUserNotificationCenter.current()
            userNotificationCenter.delegate = self
            userNotificationCenter.requestAuthorization(options: [.alert, .badge, .sound], completionHandler: callback)
        } else {
            let settings: UIUserNotificationSettings = UIUserNotificationSettings(types: [.alert, .badge, .sound], categories: nil)
            let app: UIApplication = UIApplication.shared
            app.registerUserNotificationSettings(settings)
        }
        return
    }
    @objc
    open func tokenRefreshNotification(_ notification: Notification) -> Void {
        let firInstance: InstanceID = InstanceID.instanceID()
        let messaging: Messaging = Messaging.messaging()
        let fcmToken: String? = firInstance.token()
        if (fcmToken == nil) {
            NSLog("fcmToken is null")
            return
        }
        func callback(error: Error?) -> Void {
            if (error != nil) {
                NSLog(error.debugDescription)
            } else {
                NSLog("success connect to fcm")
            }
            return
        }
        messaging.disconnect()
        messaging.connect(handler: callback)
        NSLog("fcmToken::" + firInstance.token()!)
        return
    }
    public func applicationReceivedRemoteMessage(_ remoteMessage: MessagingRemoteMessage) -> Void {
        NSLog("Received data message: \(remoteMessage.appData)")
        return
    }
    open func reset() -> Void {
        let app: UIApplication = UIApplication.shared
        app.applicationIconBadgeNumber = 0
        return
    }
    @available(iOS 10.0, *)
    public func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
        NSLog("execute fcm notification.")
        self.reset()
        completionHandler([.alert, .badge, .sound])
        return
    }
    @available(iOS 10.0, *)
    public func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Swift.Void) {
        NSLog("press fcm notification message in notify area.")
        self.reset()
        completionHandler()
        return
    }
}
